### DISCUSSION TITLE
Sending to all connected websocket clients

### DISCUSSION THREAD

**Discussion by egnor** [cesanta/mongoose#2795] [Ask questions, get answers] [answered]
I'd like to be able to broadcast a message to all websocket clients currently connected. I'm thinking this would be something like this

```
for (struct mg_connection *conn = mgr->conns; conn != NULL; conn = conn->next) {
  if (conn->is_websocket && !conn->is_draining && !conn->is_closing) {
    mg_ws_send(conn, blah blah blah);
  }
}
```

is that... reasonable? I feel like I'm mucking around in undocumented innards here. Any gotchas?

**Comment by scaprile** [comment, marked as answer]
Please see our [documentation](https://mongoose.ws/documentation/),  and follow the guidelines in our [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples).
There are several tutorials doing that.
https://mongoose.ws/documentation/tutorials/core/timers/#websocket-broadcast
https://github.com/cesanta/mongoose/blob/793171e9e0a36a5343313b53b98cf481ae2bf17f/tutorials/core/multi-threaded-12m/main.c#L69-L76

**Comment by egnor** [comment]
Ah I missed that thank you!! Maybe link that in the "WebSockets" section of the tutorials list, I didn't think to look in the "Timers" section!

**Reply by scaprile** [reply]
That is not a websockets specific action either, is just a way of doing things periodically or on demand, and timers are used to do things periodically and the second example does it on demand.

**Comment by egnor** [comment]
Yeah I was just thinking it could be linked, it's clearly a timer tutorial mainly. For websockets specifically "send to all" is a common thing one might look for within the websocket interface so a breadcrumb "also see this one" link would be handy. (Your system has the ability to "send to all that you've tagged a certain way" which is even better.)

Just a mild suggestion, thanks for the excellent code-- probably the most well organized straight-C protocol stack implementation I've ever seen, I have no idea how you managed to make it so elegant when most are so encrusted with pseudoabstractions. Kudos!

**Reply by scaprile** [reply]
I take it, will do.
WRT the code, thanks !

**Comment by Juggler-5** [comment]
Hello, Sergio. Having similar needs and read the tutorials, but.. My situation is like : Open Browser#1, establish ws logger. Data pushed to B#1. Then open B#2, establish same ws, data falling then to B#2, not to B#1 anymore - although it is still open, no connecten reset occurs, keep-alive messages (Wireshark) are coming from B#1, but no ws data(!). Then final step - closing B#2, and ws logs are neatly return to B#1. Without re-triggering ws connection at all. The only difference from https://mongoose.ws/documentation/tutorials/core/timers/#websocket-broadcast is using mg_ws_printf() instead of  mg_ws_send().

**Comment by egnor** [comment]
Can you post a complete repro case @Juggler-5 ?

**Reply by Juggler-5** [reply]
Well, a complete code of mine to build&run looks too heavy, but I can spare any details in general and particular, of course, if you would like.  But now the sceleton is 100% like this: https://github.com/cesanta/mongoose/blob/d431f4b/examples/timers/main.c#L32-L40. I think (not tried yet) that building this code can also produce similar things - ws data is sending to LAST active connection, and returns back to initial one when this new connection closed. Without, I admit, resetting/closing/reopening OLD connection again.

**Comment by scaprile** [comment]
It is your code that should do this
```
static void timer_fn(void *arg) {
  struct mg_mgr *mgr = (struct mg_mgr *) arg;
  // Broadcast "hi" message to all connected websocket clients.
  // Traverse over all connections
  for (struct mg_connection *c = mgr->conns; c != NULL; c = c->next) {
    // Send only to marked connections
    if (c->data[0] == 'W') mg_ws_printf(c, WEBSOCKET_OP_TEXT, "hi");
  }
}
```
and it is your code that should mark `wc->data[0] = 'W'`. What else can I tell you ? I know "Works for me" is not very helpful but besides checking you are correctly marking your connections, I don't know.

<img width="1685" height="1021" alt="Screenshot from 2025-12-15 14-55-43" src="https://github.com/user-attachments/assets/42cdfb78-c64c-4974-93b3-eb29c31038fb" />

**Reply by scaprile** [reply]
Remember Mongoose is not thread-safe, do you ? The same code snippet works for me, why wouldn't run for you ?

I opened one connection, saw the text, opened a second one, saw the text, then opened a third one... That code "Traverses over all connections"
